### PR TITLE
Fix RUBY-1647 Reset connection pool on non-timeout network errors

### PR DESCRIPTION
**Pull request by p-mongo**

### CODE REVIEW

File: spec/integration/sdam_error_handling_spec.rb [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  before(:all) do
+    ClientRegistry.instance.close_all_clients
+  end
+
+  describe 'when there is an error during an operation' do
+    let(:client) { authorized_client }
+
+    before do
+      wait_for_all_servers(client.cluster)
+      # we also need a connection to the primary so that our error
```
Comment (saghm) [inline spec/integration/sdam_error_handling_spec.rb]: It's not clear to me at all what this comment has to do with the code after it; what does needing a connection to the primary have to do with pinging the server and stopping the monitors?

File: spec/support/client_registry.rb [hunk 2]
```diff
@@ -95,13 +95,19 @@ def new_global_client(name)
       )
     # Provides an authorized mongo client that retries writes.
     when 'authorized_with_retry_writes'
-      global_client('authorized').with(retry_writes: true).tap do |client|
+      global_client('authorized').with(
+        retry_writes: true,
+        server_selection_timeout: 4.97,
```
Comment (saghm) [inline spec/support/client_registry.rb]: How were these server selection timeouts (4.97 and 4.99) chosen? Given how specific they are, I think it makes sense to comment why they are what they are.

File: lib/mongo/error/operation_failure.rb [hunk 3]
```diff
@@ -28,7 +29,7 @@ class OperationFailure < Error
       # @api private
       WRITE_RETRY_ERRORS = [
         {:code_name => 'InterruptedAtShutdown', :code => 11600},
-        {:code_name => 'InterruptedDueToReplStateChange', :code => 11602},
+        {:code_name => 'InterruptedDueToStepDown', :code => 11602},
```
[the remark this answers is not in the document]
Comment (p-mongo) [reply lib/mongo/error/operation_failure.rb]: See https://github.com/mongodb/specifications/pull/496.

File: spec/support/client_registry.rb (hunk 2 above)
Comment (p-mongo) [reply spec/support/client_registry.rb]: They are numbers which can be traced to where in the code they are defined, as opposed to a number like `5` which is hard to grep for. The fractions chosen are arbitrary.

File: lib/mongo/error/operation_failure.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (saghm) [reply lib/mongo/error/operation_failure.rb]: Ah, makes sense

File: spec/integration/sdam_error_handling_spec.rb (hunk 1 above)
Comment (p-mongo) [reply spec/integration/sdam_error_handling_spec.rb]: Ping establishes a connection to the primary, the comment goes with the ping. Stopping the monitors is separate. Ultimately this is all done to ensure assertions are performed on the right connections (non-monitoring vs monitoring).

**Comment by p-mongo** [comment]
@saghm Anything else for this PR?

**Pull request merged**